File: MCP-API/qdrant/src/qdrant_connector/upstream_mcp.py
from __future__ import annotations

import os
from typing import Any
from mcp import ClientSession, StdioServerParameters
from mcp.client.stdio import stdio_client
from .config import Config


class OfficialQdrantMcp:
    def __init__(self, config: Config):
        self.config = config

    def _params(self) -> StdioServerParameters:
        env = dict(os.environ)
        env["QDRANT_URL"] = self.config.qdrant_url
        if self.config.api_key:
            env["QDRANT_API_KEY"] = self.config.api_key
        if self.config.default_collection:
            env["COLLECTION_NAME"] = self.config.default_collection
        return StdioServerParameters(command=self.config.uvx_command, args=["mcp-server-qdrant"], env=env)

    async def call(self, tool: str, arguments: dict[str, Any]) -> Any:
        async with stdio_client(self._params()) as (read, write):
            async with ClientSession(read, write) as session:
                await session.initialize()
                tools = {t.name for t in (await session.list_tools()).tools}
                if tool not in tools:
                    raise LookupError(f"Official Qdrant MCP tool not available: {tool}")
                result = await session.call_tool(tool, arguments)
                if result.isError:
                    raise RuntimeError(f"Official Qdrant MCP error: {result.content}")
                return [getattr(item, "text", str(item)) for item in result.content]

File: MCP-API/okta/src/upstream-mcp.ts
import { Client } from '@modelcontextprotocol/client';
import { StdioClientTransport } from '@modelcontextprotocol/client/stdio';
import type { OktaConfig } from './config.js';

export class OktaUpstreamMcp {
  private client?: Client;
  private transport?: StdioClientTransport;
  private availableTools = new Set<string>();

  constructor(private readonly config: OktaConfig) {}

  async connect(): Promise<void> {
    if (!this.config.mcpEnabled || this.client) return;
    const env: Record<string, string> = {
      ...(Object.fromEntries(Object.entries(process.env).filter(([, value]) => typeof value === 'string')) as Record<string, string>),
      OKTA_ORG_URL: this.config.orgUrl,
      ...(this.config.mcpClientId ? { OKTA_CLIENT_ID: this.config.mcpClientId } : {}),
      ...(this.config.mcpScopes ? { OKTA_SCOPES: this.config.mcpScopes } : {}),
      ...(this.config.mcpPrivateKey ? { OKTA_PRIVATE_KEY: this.config.mcpPrivateKey } : {}),
      ...(this.config.mcpKeyId ? { OKTA_KEY_ID: this.config.mcpKeyId } : {})
    };
    this.transport = new StdioClientTransport({
      command: this.config.mcpCommand,
      args: this.config.mcpArgs,
      cwd: this.config.mcpDirectory,
      env,
      maxBufferSize: 10 * 1024 * 1024
    });
    this.client = new Client({ name: 'ai-engineering-okta-router', version: '1.0.0' }, { versionNegotiation: { mode: 'auto' } });
    await this.client.connect(this.transport);
    const listed = await this.client.listTools();
    this.availableTools = new Set(listed.tools.map((tool) => tool.name));
  }

  hasTool(name: string): boolean { return this.availableTools.has(name); }

  async call(name: string, args: Record<string, unknown>): Promise<unknown> {
    if (!this.client) await this.connect();
    if (!this.client || !this.hasTool(name)) throw new Error(`Official Okta MCP tool unavailable: ${name}`);
    const result = await this.client.callTool({ name, arguments: args });
    if ('structuredContent' in result && result.structuredContent !== undefined) return result.structuredContent;
    const text = result.content?.find((item) => item.type === 'text');
    if (text && text.type === 'text') {
      try { return JSON.parse(text.text) as unknown; } catch { return { text: text.text }; }
    }
    return result.content ?? [];
  }

  async close(): Promise<void> {
    await this.client?.close();
    this.client = undefined;
    this.transport = undefined;
    this.availableTools.clear();
  }
}
